_posts/2019-08-01-my-own-summer-neural-network.md

```diff
@@ -0,0 +1,321 @@
+
+You might ask why I didn't just pick a simple game from the internet? It's because I wanted to keep the machine learning part as simple as possible, and thought that the easiest way to achieve this is to have full control of the game mechanics.
+
+## The Game - Endless Flighter
+
+I poured another cup of joe, and rolled up some [Parcel](https://parceljs.org/) and [Three.js](https://threejs.org/). After couple of hours I had [Endless Flighter](https://github.com/juhofriman/endless-flighter) all set up and running on my browser.
+
+![Endless Flighter Human](/img/endless-flighter-human.gif)
+
+Almost as simple as possible game in which you can steer left or right and you should avoid obstacles. I did not even implement collision detection to my game, because I realised that it is not actually needed and it only makes things more complex. This way my learning data collection and testing the model is much more straight forward.
+
+## The AI Infrastructure
+
+The rough idea I learned from those Python Plays GTA videos is this:
+
+1. Implement capturing mechanism that takes screenshots, preprocesses them and registers pressed keys at the time of screenshot
+2. Capture lots of data by flying manually
+3. Balance the data
+4. Pour data into [TensorFlow](https://www.tensorflow.org/) which creates a model from your data (this is the ???? part for me)
+5. Implement another program that captures the screen, preprocesses the image with **the same exact mechanism** as in capturing stage, and then ask neural network for the prediction for the image. Then press buttons as predicted 👌
+6. You have an autonomous agent 🤖
+
+I shamelessly copied most of the code by **sentdex** almost directly and just tweaked things to fit my needs and ideas.
+
+![Rough Architecture](/img/endless-flighter-architecture.png)
+
+All the code used in my experiment is available [here](https://github.com/juhofriman/endless-flighter-ai).
+
+### 1. Capturing the data
+
+So the first thing to do is to implement a program, which captures whole lotta screenshots of manual flying and tracks keypresses.
+
+This is actually pretty easy to do. Just place your game to a certain position in your screen and use `ImageGrab` or similar to grab that image.
+
+Then you probably need to grayscale the image because colour images are huge compared to grayscale images. If you have more complex game, you probably should also apply edge detection and other image processing techniques to images in order to get them more simple for the neural network. Grayscaling seemed to be enough for my simple game. It's also important to make learning data as small as possible, so don't use 600x600 images but instead scale images down to 60x60 or so - it should be enough for the neural network to figure things out.
```

Comment:
```suggestion
Then you probably need to grayscale the image because colour images are huge compared to grayscale images. If you have more complex game, you probably should also apply edge detection and other image processing techniques to the images in order to render them more simple for the neural network. Grayscaling seemed to be enough for my simple game. It's also important to make the learning data as small as possible, so don't use 600x600 images but instead scale images down to 60x60 or so - it should be enough for the neural network to figure things out.
```